---
layout: post
author: Raúl Huertas
title: Catch2
tags: [C++, Test, TDD]
date: 2020-11-22
comments: false
---

## Catch2

Catch2 es una librería para hacer tests en C++.

Sí, sí, hay como varios cientos de librerías para hacer tests en C/C++, así que... ¿porqué vamos a dedicar nuestro tiempo a otra librería más?

Pues por varios motivos.

Uno de ellos es la facilidad para empezar a usarla. Mientras que otros frameworks exigen crear previamente una (o varias) librerías que deben enlazarse con tus ejecutables de test, Catch está compuesto unicamente por un fichero de cabecera.

Otro motivo es que es realmente cómoda de usar en comparación con las alternativas. Por poner un ejemplo, Google Test requiere conocimiento de un buen montón de macros (ASSERT_NE, ASSERT_GT, REQUIRE_LT, ...). En Catch2 hay una única macro: REQUIRE.

Testear ya es bastante complicado y normalmente poco gratificante (excepto que hayas venido a la luz de TDD). Usar cualquier librería que exija una carga cognitiva media o elevada seguramente provoque que no quieras hacer tests.

Pero hay muchos otros motivos que espero al menos se intuyan después de leer este post.


## Basta de charla, ¡enséñame el código!

Este es el ejemplo mínimo de test con Catch2:

``` cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>

TEST_CASE() {}
```

Yo he decidido descargarme [desde la rama principal de Catch2](https://github.com/catchorg/Catch2/tree/v2.x), el fichero [catch.hpp](https://github.com/catchorg/Catch2/releases/download/v2.13.3/catch.hpp) y dejarlo en un subdirectorio catch2.

Ordenado que soy, ya ves.

Y por eso lo incluyo como lo incluyo.

Compilo:
```
g++ -o p001 -I. p001.cpp
```

Y ejecuto:
```
$ ./p001
===============================================================================
test cases: 1 | 1 passed
assertions: - none -
```

### ¿CATCH_CONFIG_MAIN?

Esta macro, que hay que definir antes de incluir catch.hpp, le pide a catch que nos provea de una función main().

(El lector avezado se dará cuenta de que esta es una macro. Y no es la macro "única" que había prometido que Catch usaba. Bueno, son solo dos, no es para tanto)

Este main ejecutará todos los casos de test... y mucho más.

Ejecuta el test anterior con "--help" y verás:

```
$ ./p001 --help

Catch v2.13.3
usage:
  p001 [<test name|pattern|tags> ... ] options

where options are:
  -?, -h, --help                            display usage information
  -l, --list-tests                          list all/matching test cases
  -t, --list-tags                           list all/matching tags
  -s, --success                             include successful tests in
                                            output
  -b, --break                               break into debugger on failure
  -e, --nothrow                             skip exception tests
  -i, --invisibles                          show invisibles (tabs, newlines)
  -o, --out <filename>                      output filename
  -r, --reporter <name>                     reporter to use (defaults to
                                            console)
  -n, --name <name>                         suite name
  -a, --abort                               abort at first failure
  -x, --abortx <no. failures>               abort after x failures                                                                         
  -w, --warn <warning name>                 enable warnings                                                                                
  -d, --durations <yes|no>                  show test durations                                                                            
  -D, --min-duration <seconds>              show test durations for tests                                                                  
                                            taking at least the given number                                                               
                                            of seconds                                                                                     
  -f, --input-file <filename>               load test names to run from a                                                                  
                                            file                                                                                           
  -#, --filenames-as-tags                   adds a tag for the filename                                                                    
  -c, --section <section name>              specify section to run                                                                         
  -v, --verbosity <quiet|normal|high>       set output verbosity                                                                           
  --list-test-names-only                    list all/matching test cases                                                                   
                                            names only                                                                                     
  --list-reporters                          list all reporters                                                                             
  --order <decl|lex|rand>                   test case order (defaults to
                                            decl)
  --rng-seed <'time'|number>                set a specific seed for random
                                            numbers
  --use-colour <yes|no>                     should output be colourised
  --libidentify                             report name and version according
                                            to libidentify standard
  --wait-for-keypress <never|start|exit     waits for a keypress before
  |both>                                    exiting
  --benchmark-samples <samples>             number of samples to collect
                                            (default: 100)
  --benchmark-resamples <resamples>         number of resamples for the
                                            bootstrap (default: 100000)
  --benchmark-confidence-interval           confidence interval for the
  <confidence interval>                     bootstrap (between 0 and 1,
                                            default: 0.95)
  --benchmark-no-analysis                   perform only measurements; do not
                                            perform any analysis
  --benchmark-warmup-time                   amount of time in milliseconds
  <benchmarkWarmupTime>                     spent on warming up each test
                                            (default: 100)

For more detailed usage please see the project docs
```

Pues no está mal para tres líneas de código, ¿verdad?

### TEST_CASE

TEST_CASE es otra de las macros principales de Catch2.

Sí, había dicho que solo había una y ya van tres, pero era un clickbait. Igual gracias a ello sigues leyendo, ¿no? Pues ya me ha servido.

El caso es que esta macro crea un caso de test. Si se usa sin parámetros el test no tendrá nombre, pero le podemos dar nombre:

``` cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>

TEST_CASE("We will rock you") {

}
```

Podemos comparar las salidas de este nuevo ejecutable p002 con el anterior p001 para entender para qué usa Catch2 estos nombres:
```
$ ./p001
===============================================================================
test cases: 1 | 1 passed
assertions: - none -

$ ./p002
===============================================================================
test cases: 1 | 1 passed
assertions: - none -
```

Pues sí, yo tampoco veo la diferencia.

Prueba a ejecutarlo pidiendo que te liste el tiempo de ejecución de cada test case:
```
$ ./p001 -d yes
0.000 s: Anonymous test case 1
===============================================================================
test cases: 1 | 1 passed
assertions: - none -

$ ./p002 -d yes
0.000 s: We will rock you
===============================================================================
test cases: 1 | 1 passed
assertions: - none -

```

Normalmente Catch2 no imprime nada. Pero si falla un test o en respuesta a algún parámetro pasado (como el "-d" anterior), entonces mostrará el nombre del test. 

Mientras que en el primer caso se ha inventado un nombre "Anonymous test case 1", en cuanto le demos nombre Catch lo usará cuando tenga algo relevante que decir sobre el mismo.

## Testeando cosas

Si tu paciencia te ha dado para llegar aquí, además de mi pésame te voy a regalar cómo poner aserciones en Catch2. Al fin y al cabo, lo visto hasta ahora no es que sirva de mucho...

``` cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>

TEST_CASE("We will rock you") {
    REQUIRE( 3*7 == 20 );
}
```

REQUIRE es _la_ macro de Catch2. No solo te evalúa la sentencia pasada, además te la mostrará caso de que la aserción no se cumpla (es decir, se evalúe a false), pero, sobretodo, la "expandirá".

Veamos:
```
$ ./p003

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
p003 is a Catch v2.13.3 host application.
Run with -? for options

-------------------------------------------------------------------------------
We will rock you
-------------------------------------------------------------------------------
p003.cpp:4
...............................................................................

p003.cpp:5: FAILED:
  REQUIRE( 3*7 == 20 )
with expansion:
  21 == 20

===============================================================================
test cases: 1 | 1 failed
assertions: 1 | 1 failed
```

La parte que hace que Catch2 sea, en mi opinión, superior a otras alternativas son estas dos líneas:
```
with expansion:
  21 == 20
```

Es por conseguir algo similar que otros frameworks de tests tienen tantas macros.

En Google Test, por ejemplo, deberíamos usar:
```
ASSERT_EQ(3*7, 20);
```
para conseguir un mensaje igual de informativo. Porque si usamos el más genérico
```
ASSERT_TRUE(3*7 == 20);
```
el mensaje que muestra es algo del estilo
```
Assertion failed: 3*7 == 20
Should be true but it is false.
```

Para este ejemplo tan simple igual no parece tan importante, pero en un ejemplo real estaremos llamando a una o varias funciones, y el mostrar la expansión es imprescindible para entender el error:

``` cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>


int f(int a, int b) {
    return a*b;
}

int g(int a) {
    return a*10;
}

TEST_CASE("We will rock you") {
    REQUIRE( f(3,7) == g(2) );
}
```

Muestra:
```
p004.cpp:14: FAILED:
  REQUIRE( f(3,7) == g(2) )
with expansion:
  21 == 20
```

Bah, te está gustando, ¿eh?

### Expresion templates

¿Cómo consigue Catch2 hacer eso?

Pues porque usa magia. Le dan el nombre de "expression templates" pero en el fondo es magia.

Eso sí, es una magia no tan poderosa. Antes de que te vuelvas loco a probar cosas raras, mira qué pasa si simplemente aumentamos un poco la complejidad de la sentencia REQUIRE:

``` cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>

TEST_CASE("We will rock you") {
    REQUIRE( 3*7 == 20 || 10 < 5 );
}
```

Mensajito al compilar:
```
In file included from p005.cpp:2:0:
./catch2/catch.hpp: In instantiation of ‘const Catch::BinaryExpr<LhsT, const RhsT&> Catch::BinaryExpr<LhsT, RhsT>::operator||(T) const [with T = bool; LhsT = const int&; RhsT = const int&]’:
p005.cpp:5:5:   required from here
./catch2/catch.hpp:2246:13: error: static assertion failed: chained comparisons are not supported inside assertions, wrap the expression inside parentheses, or decompose it
             static_assert(always_false<T>::value,
             ^~~~~~~~~~~~~
```

Para lo que es C++, yo diría que hasta es un error agradable de leer.

Si hacemos lo que nos pide, envolver con paréntesis la aserción, compilará, pero la expansión que hará cuando la aserción no se cumpla dejará bastante que desear:
```
p005.cpp:5: FAILED:
  REQUIRE( (3*7 == 20 || 10 < 5) )
with expansion:
  false
```

¿Es esto un problema?

Pues para mí absolutamente ninguno.

En Google Test, por comparar con frameworks que usan multitud de macros, no existen macros para expresiones tan complejas. Existe ASSERT_EQ (que busca la igualdad de sus dos parámetros) y ASSERT_LT (que busca que el primer parámetro sea menor que el segundo).

Pero no existe una macro ASSERT_EQ_OR_LT que haga todo lo que hemos querido poner en el REQUIRE.

Yo jamás he necesitado poner expresiones tan complejas en un test. Si eres de los que te gustaría poder escribir este tipo de cosas, háztelo ver, de verdad.

## Tags

Pasemos a otro tema.

El problema de empezar a escribir tests es que engancha. Y antes de que lo veas venir, tienes cientos de tests en tus proyectos.

Así que conviene tenerlos organizados. Y para eso se usa el segundo parámetro de la macro TEST_CASE:
```
TEST_CASE( "A", "[widget]"         ) { /* ... */ }
TEST_CASE( "B", "[widget]"         ) { /* ... */ }
TEST_CASE( "C", "[gadget]"         ) { /* ... */ }
TEST_CASE( "D", "[widget][gadget]" ) { /* ... */ }
```

Como verás los tags se pasan en el segundo parámetro. Deben usar corchetes para saber donde termina uno y empieza otro.

Y se pueden usar cuando queremos ejecutar solo un subconjunto de todos los tests.

El primer parámetro que se pasa a los ejecutables Catch2 es una cadena de texto que permite filtrar los tests a ejecutar mediante tags.

En el ejemplo anterior:
* poniendo "[widget]" se ejecutarán los tests A, B y D.
* poniendo "[gadget]" se ejecutarán C y D.
* poniendo "[gadget][widget]" se exige que tenga los dos (es un AND) así que solo se ejecutará D.
* poniendo "[gadget],[widget]" se exige que tenga alguno de los dos (es un OR) así que se ejecutarán todos.

Es un sistema bastante potente y creo que suficiente para todos los requisitos racionales.

Pero entiendo que siempre hay gente rara por ahí fuera (sí, te estoy mirando a tí), que quiere más.

Pues resulta que también se pueden usar especie de expresiones regulares para filtrar los tags:
```
thisTestOnly            Matches the test case called, 'thisTestOnly'
"this test only"        Matches the test case called, 'this test only'
these*                  Matches all cases starting with 'these'
exclude:notThis         Matches all tests except, 'notThis'
~notThis                Matches all tests except, 'notThis'
~*private*              Matches all tests except those that contain 'private'
a* ~ab* abc             Matches all tests that start with 'a', except those that
                        start with 'ab', except 'abc', which is included
-## [#somefile]          Matches all tests from the file 'somefile.cpp'
```

Francamente, no sé para qué alguien puede querer estas cosas, pero si te ves en esa situación te vas a [la documentación oficial](https://github.com/catchorg/Catch2/blob/devel/docs/command-line.md#specifying-which-tests-to-run) y ya nos cuentas en un post, ¿vale?

## Secciones

Esta es otra característica diferenciadora de Catch2.

Las secciones son una idea muy interesante que permite olvidarte de los dichosos "test fixtures" de otros frameworks de test.

``` cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>

TEST_CASE("Usemos secciones") {
    int una_variable{10};

    SECTION("Podemos cambiar el valor de la variable") {
        ++una_variable;
        REQUIRE(una_variable == 11);
    }

    SECTION("Y sin embargo no hay fuga de modificaciones") {
        REQUIRE(una_variable == 10);
    }
}
```

Fíjate que las secciones son independientes una de otra... pero comparten el código previo a ellas.

La declaración de "una_variable" para ser usada en dos secciones requeriría un test fixture en otros frameworks de test. Sin embargo en Catch2 el código queda completamente natural.

Lo interesante es que catch2 ejecuta el test case desde el principio para cada SECTION. Es decir, el constructor de una_variable (si lo tuviera) se ejecutaría dos veces en el ejemplo anterior.

¿Cómo lo consigue? Pues ya sabes, magia. Pero esta vez es bastante poderosa, porque no he conseguido confundirla nunca.

Una de las cosas más interesantes es que podemos poner código entre secciones. Mediante test fixtures tendríamos que llevarnos ese código a los constructores. Si no sabes de qué te hablo, da igual, simplemente mira cómo de forma natural podemos hacer cosas muy interesantes con secciones:

``` cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>

TEST_CASE("Usemos secciones") {
    int una_variable{10};
    una_variable *= 2;

    SECTION("Por ahora una_variable vale 20") {
        ++una_variable;
        REQUIRE(una_variable == 21);
    }

    una_variable -=5;

    SECTION("Y ahora vale 15") {
        ++una_variable;
        REQUIRE(una_variable == 16);
    }

    SECTION("Y sigue valiendo 15") {
        REQUIRE(una_variable == 15);
    }
}
```

Incluso se pueden poner secciones dentro de secciones.

Que no se te vaya la olla, que luego hay gente que tiene que leer tus tests y entenderlos.

## Estilo BDD

Me incluyo entre los chalados que les mola leer los tests en plan Given-When-Then.

Pues bien, en Catch2 han creado unas macros para poder escribir los tests así (si no he perdido la cuenta con estas ya son 6 las macros usadas, para que luego digas que no te miento):

```
SCENARIO( "vectors can be sized and resized", "[vector]" ) {

    GIVEN( "A vector with some items" ) {
        std::vector<int> v( 5 );

        REQUIRE( v.size() == 5 );
        REQUIRE( v.capacity() >= 5 );

        WHEN( "the size is increased" ) {
            v.resize( 10 );

            THEN( "the size and capacity change" ) {
                REQUIRE( v.size() == 10 );
                REQUIRE( v.capacity() >= 10 );
            }
        }
    }
}
```

No son las únicas, también encontrarás AND_GIVEN, AND_WHEN y similares. La idea es que cuando una aserción falle, te mostrará algo así:
```
Scenario: vectors can be sized and resized
     Given: A vector with some items
      When: the size is increased
      Then: the size and capacity change
```

En mi opinión, los tests quedan más legibles y enfocados si se usan de esta forma. Pero, oye, si no te gusta, pues utilizas simplemente REQUIRE y ya está.

## ¡Tarda mucho en compilar!

No me queda la más mínima duda de que habrás ido copiando todos los ejemplos del post, compilandolos y ejecutándolos. Por supuesto. Claro.

Y te habrás dado cuenta de que tardan bastante en compilar. Francamente, es un rollazo.

Pero es el precio a pagar por tener todo en un único fichero de cabecera.

Sin embargo, hay una solución.

Te creas un fichero main.cpp (puedes cambiar el nombre) con este contenido:

```
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>
```

Fíjate que en este fichero dejamos CATCH_CONFIG_MAIN.

Y, por otro lado, te creas un fichero cpp por cada conjunto significativo de tests. Por ejemplo, p008.cpp:

```
#include <catch2/catch.hpp>

TEST_CASE() {}
```
Fíjate que en estos ficheros omitimos CATCH_CONFIG_MAIN.

Ahora los compilas por separado y los enlazas juntos:
```
$ g++ -I . -c main.cpp
$ g++ -I . -c p008.cpp 
$ g++ -o p008 main.o p008.o
```

(Por supuesto, usa CMake, pero no te voy a dar ahora un curso de eso, ¿no?)

Compilar main.cpp seguirá llevando bastante tiempo, pero no hace falta que lo recompiles cada dos por tres, porque nunca lo cambiarás.

Estarás cambiando el resto de ficheros de tests. Pero esos tardan muy muy poco en recompilarse gracias a la omisión de CATCH_CONFIG_MAIN.

## Matchers

Con la macro REQUIRE deberías cubrir el 80% de tus necesidades. Por no decir el 99%.

Sin embargo hay veces que las expansiones de las aserciones dejan bastante que desear.

Pongamos que quieres escribir una aserción que compruebe que la cadena de texto devuelta por una función contiene una subcadena dada:

```
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>
#include <string>

std::string f() {
    return "Tres trostes tigres";
}

TEST_CASE("Los tigres deberían estar tristes") {
    REQUIRE( f().find("tristes") != std::string::npos );
}
```

Ejecutemos:
```
$ ./p009

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
p009 is a Catch v2.13.3 host application.
Run with -? for options

-------------------------------------------------------------------------------
Los tigres deberían estar tristes
-------------------------------------------------------------------------------
p009.cpp:9
...............................................................................

p009.cpp:10: FAILED:
  REQUIRE( f().find("tristes") != std::string::npos )
with expansion:
  18446744073709551615 (0xffffffffffffffff)
  !=
  18446744073709551615 (0xffffffffffffffff)

===============================================================================
test cases: 1 | 1 failed
assertions: 1 | 1 failed
```

Vale, vale. Supongo que si lo piensas un poco tiene sentido, pero... joer, ¡vaya porquería de mensaje de error!

Lo que necesitamos es una forma de especificarle a Catch el mensaje que queremos que se imprima cuando la aserción falla.

Y parece una cosa sencilla, pero no lo es tanto. 

En Catch2 podemos hacer uso de una funcionalidad para solucionar este tema: los Matchers.

Según la documentación oficial, "los Matchers son una forma alternativa de escribir aserciones, útiles para tests donde se trabaje con tipos complejos o donde sea necesario hacer aserciones sobre propiedades más complejas."

Pero yo los voy a usar simplemente para que el mensaje de error quede más legible.

Para ello en vez de usar REQUIRE debemos usar REQUIRE_THAT (sí, sí, otra macro más... ¿dejamos ya el temita?)

Esta macro recibe dos parámetros, el primero es el valor que queremos testear y el segundo un matcher que queremos que sea cierto:
```
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>
#include <string>

using Catch::Matchers::Contains;

std::string f() {
    return "Tres trostes tigres";
}

TEST_CASE("Los tigres deberían estar tristes") {
    REQUIRE_THAT( f(), Contains("tristes"));
}
```

Ahora el error es mucho más claro:
```
$ ./p010

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
p010 is a Catch v2.13.3 host application.
Run with -? for options

-------------------------------------------------------------------------------
Los tigres deberían estar tristes
-------------------------------------------------------------------------------
p010.cpp:11
...............................................................................

p010.cpp:12: FAILED:
  REQUIRE_THAT( f(), Contains("tristes") )
with expansion:
  "Tres trostes tigres" contains: "tristes"

===============================================================================
test cases: 1 | 1 failed
assertions: 1 | 1 failed
```

Mucho mejor.

Pero es que además los podemos combinar:
```
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>
#include <string>

using Catch::Matchers::EndsWith;
using Catch::Matchers::Contains;

std::string f() {
    return "Tres trostes tigres";
}

TEST_CASE("Los tigres deberían estar tristes") {
    REQUIRE_THAT( f(),
              EndsWith("tigres") && Contains("tristes"));
}
```

Ejecutando:
```
p011.cpp:14: FAILED:
  REQUIRE_THAT( f(), EndsWith("tigres") && Contains("tristes") )
with expansion:
  "Tres trostes tigres" ( ends with: "tigres" and contains: "tristes" )
```
 

Como me indica un gran conocedor del C++ (gracias @Manu343726), hasta queda más 
legible si en el test usamos "and" en vez de "&&":
```
TEST_CASE("Los tigres deberían estar tristes") {
    REQUIRE_THAT( f(),
              EndsWith("tigres") and Contains("tristes"));
}
```
¡Y así reivindicamos a esas grandes keywords desconocidas: 
[and](https://en.cppreference.com/w/cpp/keyword/and) y [or](https://en.cppreference.com/w/cpp/keyword/or).

Si estás pensando en usar Catch2 de forma seria mi recomendación es que conozcas los Matchers. Es mi opinión que no deben usarse para todos los tests, pero son muy útiles en más de una ocasión... y no solo para que el mensaje de error quede más bonito.

## Tenemos que ir acabando

Mi Community Manager me está diciendo que el post está quedando demasiado largo.

Pero es que me estoy dejando muchas cosas en el tintero. Así, a bote pronto:
- Clara, the Command Line parser
- Generadores
- Instanciación automática de templates
- Y, en general, un buen montón de macros

En fin, supongo que si recibo muchos comentarios a favor de hacer otro post sobre esos temas lo terminaré haciendo...

Como precaución he desactivado los comentarios, que tampoco es cosa de currar por currar...
